FlowCrypt/src/main/java/com/flowcrypt/email/ui/adapter/RecipientChipRecyclerViewAdapter.kt

```diff
@@ -0,0 +1,311 @@
+ */
+class RecipientChipRecyclerViewAdapter(
+  val recipientType: Message.RecipientType,
+  private val onChipsListener: OnChipsListener
+) : ListAdapter<RecipientChipRecyclerViewAdapter.Item,
+    RecipientChipRecyclerViewAdapter.BaseViewHolder>(DIFF_CALLBACK) {
+  private var addViewHolder: AddViewHolder? = null
+
+  var resetTypedText = false
+    set(value) {
+      field = value
+      if (value) {
+        addViewHolder?.binding?.editTextEmailAddress?.text = null
+      }
+    }
+
+  override fun onCreateViewHolder(
+    parent: ViewGroup,
+    viewType: Int
+  ): RecipientChipRecyclerViewAdapter.BaseViewHolder {
+    return when (viewType) {
+      ADD -> {
+        addViewHolder = AddViewHolder(
+          LayoutInflater.from(parent.context)
+            .inflate(R.layout.compose_add_recipient_item, parent, false)
+        )
+        requireNotNull(addViewHolder)
+      }
+
+      MORE -> MoreViewHolder(
+        LayoutInflater.from(parent.context)
+          .inflate(R.layout.chip_more_item, parent, false)
+      )
+
+      else -> ChipViewHolder(
+        LayoutInflater.from(parent.context)
+          .inflate(R.layout.chip_recipient_item, parent, false)
+      )
+    }
+  }
+
+  override fun onBindViewHolder(
+    holder: RecipientChipRecyclerViewAdapter.BaseViewHolder,
+    position: Int
+  ) {
+    when (holder) {
+      is AddViewHolder -> holder.bind()
+      is ChipViewHolder -> holder.bind(getItem(position).itemData as RecipientInfo)
+      is MoreViewHolder -> holder.bind(getItem(position).itemData as ItemData.More)
+    }
+  }
+
+  override fun getItemViewType(position: Int): Int {
+    return getItem(position).type
+  }
+
+  fun submitList(recipients: Map<String, RecipientInfo>, isModifyingEnabled: Boolean = true) {
+    val filteredList = recipients.values
+      .take(if (hasInputFocus()) recipients.size else MAX_VISIBLE_ITEMS_COUNT)
+    val hasUpdatingInHiddenItems = (recipients.values - filteredList.toSet()).any { it.isUpdating }
+    val recipientInfoList = filteredList
+      .map { Item(CHIP, it.copy(isModifyingEnabled = isModifyingEnabled)) }
+    val finalList = recipientInfoList.toMutableList().apply {
+      if (recipients.size > MAX_VISIBLE_ITEMS_COUNT && !hasInputFocus()) {
+        add(
+          Item(
+            MORE,
+            ItemData.More(
+              value = recipients.size - filteredList.size,
+              hasUpdatingInHiddenItems = hasUpdatingInHiddenItems
+            )
+          )
+        )
+      }
+      if (isModifyingEnabled) {
+        add(Item(ADD, ItemData.ADD))
+      }
+    }
+    submitList(finalList)
+  }
+
+  fun requestFocus() {
+    addViewHolder?.binding?.editTextEmailAddress?.requestFocus()
+  }
+
+  private fun hasInputFocus(): Boolean {
+    return addViewHolder?.binding?.editTextEmailAddress?.hasFocus() == true
+  }
+
+  abstract inner class BaseViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+    fun prepareProgressDrawable(): Drawable {
```

Comment:
can be protected?

Reply:
Yup. It can be. Or can be static. We can change it later, it's a minor improvement.